# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/Shape.cpp
# Commit: 503f3d7f7162
# Full Hash: 503f3d7f71626b73c7e72a7c6f6d58ba9c7bc523
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2021-03-06 09:33:16
# Regressor Bug: 1689413
# File Overlap Count: 1
# Description:
#   Bug 1689413 part 14 - Set the Delegate flag in EmptyShape::getInitialShape if needed. r=tcampbell
#   
#   We were relying on this flag being set in ObjectGroup::defaultNewGroup, but that's
#   going away soon.
#   
# ==============================================================================

diff -r 5986697cbefa -r 503f3d7f7162 js/src/vm/Shape.cpp
--- a/js/src/vm/Shape.cpp	Sat Mar 06 01:05:24 2021 +0000
+++ b/js/src/vm/Shape.cpp	Sat Mar 06 01:05:24 2021 +0000
@@ -1469,6 +1469,7 @@
 
   MOZ_ASSERT_IF(proto().isObject(),
                 compartment() == proto().toObject()->compartment());
+  MOZ_ASSERT_IF(proto().isObject(), proto().toObject()->isDelegate());
 
   // Windows may not appear on prototype chains.
   MOZ_ASSERT_IF(proto().isObject(), !IsWindow(proto().toObject()));
@@ -1984,6 +1985,14 @@
   MOZ_ASSERT_IF(proto.isObject(),
                 cx->isInsideCurrentCompartment(proto.toObject()));
 
+  if (proto.isObject() && !proto.toObject()->isDelegate()) {
+    RootedObject protoObj(cx, proto.toObject());
+    if (!JSObject::setDelegate(cx, protoObj)) {
+      return nullptr;
+    }
+    proto = TaggedProto(protoObj);
+  }
+
   auto& table = realm->zone()->initialShapes();
 
   using Lookup = InitialShapeEntry::Lookup;
